### PR TITLE
Resolve: API Abstraction proposal & BestLength Creator tool

### PR DESCRIPTION
**Pull request by tweak-wtf**
## Changelog Description
This PR is WIP. 
Proposes an object-oriented abstraction for the Resolve API in `hosts/resolve/api/wrappers`.
Uses abstraction to create BestLength Creator plugin.

## Additional info
Reasoning behind the abstraction:
- it should give a clear and concise interface for developers
- it should abstract away all quirks that some NLE's might bring up

## Testing notes:
1. launch Davinci Resolve using AYON Launcher
2. open up any Resolve project
3. open the Creator via `Scripts/Compo/OpenPype__Menu`
4. select `Create Best Length Timeline`

## Screenshot
image

### CODE REVIEW

File: openpype/hosts/resolve/api/wrappers/timeline.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    @property
+    def otio(self) -> otio.schema.Timeline:
+        return self.__otio
+
+    @otio.setter
+    def otio(self, val):
+        self.__otio = val
+
+    @property
+    def jotio(self) -> jotio.schema.Timeline:
```
[the remark this answers is not in the document]
Comment (tweak-wtf) [reply openpype/hosts/resolve/api/wrappers/timeline.py:71]: whoops. yeah that's my bad

**Comment by BigRoy** [comment]
Wait a second - the Resolve API is already (unlike e.g. Maya) object oriented. They are classes with methods, even if going through `PyRemoteObject`.

Also, I feel like having this be part of OpenPype isn't the best way forward. You're likely better of writing the abstraction separate as a package of its own so that if it happens to be production-proven and way better than what Resolve provides out of the box we could then look into using it.

Designing the wrapper inside OpenPype only means there's more maintenance burden on the OpenPype Resolve integration team.

Having it standalone also means that unittests can be made much more sensible since they'd only need to check that package + resolve instead of dealing with OpenPype surrounding it as well.

Having said that, I totally understand your endeavour - I once started [`fusionless` for Fusion](https://github.com/Colorbleed/fusionless) due to also having issues with Fusion and its Python API which was back then even less documented. I ended up not using it in Avalon/OpenPype integration since it required other devs to then learn _that_ API instead of relying on whatever knowledge they had of Fusion.

Basically what's happening with a wrapper like this is that you're adding an **additional** layer someone must learn instead just a replacement. Over time there'll also much likely be way more questions/knowledge on the web about Resolve's API instead of this redesigned API because there will just be that much more users.

Building a wrapper means needing to maintain it over Resolve releases as new features get added to Resolve, and deal with backward compatibility and deprecated methods over time as well (which Resolve has shown to do now and then).

Hence, I don't really see why this would be worthwhile continuing.

**Comment by BigRoy** [comment]
Also regarding that Best Length Plate creator - I'm not sure what it intends to do, but I have a feeling you're much better off designing it against the #5709 PR so efforts can be put there into making it work towards the new publisher.

But maybe @jakubjezek001 can say more about the goals there.

**Comment by tweak-wtf** [comment]
hey @BigRoy 
thanks for sharing your thoughts about this. :)
This is my first PR so I'm just starting out learning about this codebase so please bear with me.


i was just a bit fed up with not knowing what object I actually get (everything is `BlackmagicFusion.PyRemoteObject`) and what function there are to use. Also needed to calculate duration of a TimelineItem a bit differently than resolve does in order to account for linear time remaps as well as speed ramps.
but i could also create it as a pip package very much like `pydavinci` does it.

The idea behind the BestLength tool is that you define a set of Timelines and get timeline items for each of these timelines representing their longest occurrence.
For each Timeline I'm going over every occurrence of a mediapool item.
Then by iterating all occurrences i'm able to calculate the part of the MediapoolItem that will represent all usages in its "best length" with a configurable gapsize between usages. Hopefully soon even if speed changes are present since I export/import the otio of that timeline to gain access to the effects parameters.
This shall prevent having to work on duplicate shots/frames and alleviate editorial work during ingest

However, this is just an idea and something I did this weekend :)

**Comment by BigRoy** [comment]
> i was just a bit fed up with not knowing what object I actually get (everything is BlackmagicFusion.PyRemoteObject) and what function there are to use.

You might want to look into typings for Resolve if this is the case - have a look at e.g. https://github.com/austinwitherspoon/davinci-resolve-type-hints

Here is similar for Fusion: https://github.com/EmberLightVFX/BMD-Fusion-Scripting-Stubs

> Also needed to calculate duration of a TimelineItem a bit differently than resolve does in order to account for linear time remaps as well as speed ramps.

This sounds like a function by itself should be capable in for this scenario. No need to wrap the whole of Resolve.

> The idea behind the BestLength tool is that you define a set of Timelines and get timeline items for each of these timelines representing their longest occurrence.
> For each Timeline I'm going over every occurrence of a mediapool item.
> Then by iterating all occurrences i'm able to calculate the part of the MediapoolItem that will represent all usages in its "best length" with a configurable gapsize between usages. Hopefully soon even if speed changes are present since I export/import the otio of that timeline to gain access to the effects parameters.
> This shall prevent having to work on duplicate shots/frames and alleviate editorial work during ingest

Conceptually I like the idea "BestLengthTool" I mean is a very weird Creator name for that, but the concept sounds solid - from all used clips find the frames that were actually used and integrate only that (potentially allowing to add 'handles' on top of that even?) Might get complicated quickly haha - but fun prototyping.

**Comment by tweak-wtf** [comment]
Oh these typing look cool! gonna give them a try :)
I actually already have a working prototype for this as a standalone resolve script here:
https://github.com/tweak-wtf/resolve-merge-timelines

I had to use chatGPT to help me out with the last bit of the algo but i'm currently also rewriting this. Was bit happy that chatGPT at least used the same approach i was going for for the first part :D

**Comment by BigRoy** [comment]
Noticed that I was requested a review - I will hold back on that.

As stated before, the "wrapper" logic of this PR should not be in OpenPype/AYON.

I'd remove the wrappers - then whatever is tricky with the main API of Resolve, e.g. a certain time code calculation - move only that logic into `lib` and then rewrite the best length creator to use that.

Remove the custom `get_logger` logic, etc. OR clean that up to either make sense as a re-usable thing for Resolve. Likely that'd be removal?

I feel the logic of the finding the best clip length is hard to follow. It might be the lack of docstrings or comments, but I literally have no idea why `get_occurences`, `sort_occurences` and `get_best_length_for_clips` look this complex for as far as I can see all that you're looking for is the start frame and end frame of a particular clip (+ including any frames that are not there?)

I need some explanation as to **what** the idea is there?

I feel like in essence it is:

```python
start = None
end = None
for clip in clips:
    clip_start, clip_end = get_clip_range(clip)
    if start is None:
        start = clip_start
    else:
        start = min(clip_start, start)
    if end is None:
        end = clip_end 
    else:
        end = max(clip_end, end)

full_range = (start, end)
```

I'm not sure what the **gap size** intends to do - and likely it is the Gap Size that is complicating the logic there.
But if we need the actual gaps I'd maybe do:

```python
all_frames = set()
for clip in clips:
    frames = get_used_frames(clip)
    all_frames.update(frames)

# Now let's group by consecutive frames, considering the gap size
gap_size = 10
current_group = []
groups = [current_group]
ranges = []
for frame in sorted(all_frames):
    last_frame = current_group[-1] if current_group else frame
    if frame - last_frame <= gap:
        current_group.append(frame)
    else:
        # new group because gap is too large
        current_group = [frame]
        groups.append(current_group)

# `groups` now contains lists of frames that are more than 10 frames apart
```

Anyway, even then I'm still not sure _why_ we need to know those groups.

**Comment by jakubjezek001** [comment]
Here are my thoughts on the matter. I have reviewed all the notes from our workgroup meeting regarding this topic. This feature will be extremely beneficial for projects with multiple timeline variations, where certain clips may be shared across different timelines with varying edits.

Here is a typical example. We have three timelines for one commercial TV spot with VFX shots: 10s, 20s, and 30s variants of the edit. Some shots may have multiple components that need to be composited together from different plates (background, greenscreen01, greenscreen02). These components are usually layered in multiple tracks. In the vertically aligned workflow, all clips should have the same shot name and be aligned vertically above each other. Rest of the none VFX shots will be only graded and there is no need to publish them into Ayon. 
image

Just to establish used terminology: **Master Shot Timeline** (MST) is timeline generated from best lengths clips found in **Child Timelines** (CTs).

The tool should be capable of performing the following tasks:
- Finding shared clips across all timelines and tracks.
- Identifying the longest cuts among these shared clips.
- Creating a new timeline **MST** and adding the best length clips to it.
- Preserving the multi-track structure and its vertical relation to the main track.
- Be able to extend clips with defined lengths for the head and tail.
- Allowing the option to include or exclude handles (explicitly or implicitly).
- Establishing backlinks or references from the **MST** clips to **CTs** clips, potentially using markers. 
- To be able to rerun the tool multiple times during production, because adjustments must be made to the **MST** if the clip lengths in **CTs** has been changed.

The inspiration is taken from Flames `Shot Sequence` tool and concept is well described  [here](https://www.youtube.com/watch?v=WmU3bEx4Ng8). And for now lets ignore the `Source Sequence`.

Lets explain some of the mentioned features and its requirements and relation to current proposal:

#### Multiple tracks
Multiple tracks are very difficult to manage in current state of Resolve API so this will be difficult to approach with current state of this  proposal. For that reason we might be pushed to implement it via *OpenTimelineIO*, because new OTIO timeline with **MST** can be created, saved as **.otio** file and then imported back into the project. In theory this should solve the issue. This also means we will not need the wrapper @tweek-wtf is suggesting in this PR and rather focus on already created `openpype\hosts\resolve\otio\davinci_export.py`. So I agree with @BigRoy with his comment [here](https://github.com/ynput/OpenPype/pull/5904#issuecomment-1808004672), 

Since OTIO is already used in other hosts that deal with editorial work, this would also benefit them.

#### Backlinking clips
We have to think of the full production scope and be ready to react quickly with the tool to any changes on **CTs**. Ideally the information for the backlinking should be somehow easily approachable so users them selves can comfortably identify the relationships. 
Currently, I am conducting research on the potential use of subtitle clips in **MST** for storing metadata and improving the visualization of relationships.

#### Standalone tool
I suggest creating a standalone tool or script for this purpose. In DaVinci Resolve, a button can be added to the menu for easy access. The creator's function is to convert host objects into publishable instances, but in this case, we are only preparing data for further processing, which may not require publishing.
 
My conclusion on this proposal is this is nice POC we were able to demonstrate the concept similarly to this [Hiero script implementation](https://github.com/horitin/dotStudio/blob/master/Python/Startup/LongestShotMaker.py) @jrsndl was suggesting originally in his [repo here](https://github.com/jrsndl/resolve-merge-timelines). But we will need to focus more on the OTIO implementation to be able flexibly addressed above features ideally with host agnostic abstractions. 

Thank you to all who are participating in this feature, especially @jrsndl and @tweak-wtf . Your contributions have provided a great starting point for the ongoing conversation and have explored some valuable aspects that will benefit the feature.

**Comment by BigRoy** [comment]
Just to re-iterate, I totally agree that there's value to the issue it tries to solve and the feature it implements.

My main point is that the logic can be **greatly** simplified. There's no need for wrappers around the Resolve library for this feature - plus the code style used is very hard to read and convoluted. In my previous comment [here](https://github.com/ynput/OpenPype/pull/5904#issuecomment-1822900306) I tried to put forth an example of what the majority of the actual logic does in this PR in just a few lines of code - we don't need the 800+ lines of code.

What I'd also like is that maybe we develop this feature in a test-driven development instead. So basically we write a few tests with example input files and the expected output lengths.

```python

def test_simple_best_length():
    # Clip is used in two sequences without time remapping
    # and frame ranges overlap
    import_test_sequences("simple_example.otio")
    start, end = get_best_length(clip)
    assert start == 1001
    assert end == 1150

def test_best_length_unorded_non_overlapping():
    # Clip is used in three sequences with a random order
    # and two clips have overlapping ranges and one does
    # not overlap the others
    # e.g.
    #   sequence 1: 1010-1030
    #   sequence 2: 1001-1020
    #   sequence 3: 1050-1100
    import_test_sequences("unorded_non_overlapping.otio")
    start, end = get_best_length(clip)
    assert start == 1001
    assert end == 1100

```

Etc.

Then all we need for this feature is really the `get_best_length` function implemented?

---

> Preserving the multi-track structure and its vertical relation to the main track.

Sorry, how would it do that? If a clip is present in multiple sequences on different tracks, at what track should it end up?

> Be able to extend clips with defined lengths for the head and tail.

So just this?
```python
start, end = get_best_length(clip)
start -= handle_start
end += handle_end
```

**Comment by tweak-wtf** [comment]
firstly 1000 thanks for actually checking this out! It really means something to me.
i'm of @BigRoy 's opinion. it can be greatly simplified and yes let's do this test driven! :)

it's really just me writing the code i wish i had and then make sense out of resolve. most of the methods are completely redundant. however, some might be interesting since it appears to be kinda tricky to get linear or keyframed speed change data out of resolve. at least to me.
proposal is big word true, reject it please. let's build something cool together maybe. i guess the idea is clear.

anyways, happy holidays and thanks again! 😽

**Comment by jakubjezek001** [comment]
Here is an example of the basic code that solves two variations of the same video clip. This example generates the OTIO file. Please note that this example does not solve the requirement for multiple tracks, but it can be easily added later.

Just posting it here for future reference. 

```python
import opentimelineio as otio

# Load the timelines from the .otio files
timeline1 = otio.adapters.read_from_file("/path/to/10second.otio")
timeline2 = otio.adapters.read_from_file("/path/to/20second.otio")

# Initiate a dict to store the clips, key is the media_reference and value is the clip
clips_dict = dict()

# Function to add clips to dict only if they are longest version
def add_clips(clips_dict, track):
   for clip in track.each_clip():
      # If the clip's media_reference is not in clips_dict, add it
      # or if it is in the dict but the new clip duration is longer, replace it
      if clip.media_reference not in clips_dict or clip.duration() > clips_dict[clip.media_reference].duration():
            clips_dict[clip.media_reference] = clip

# Add clips from both timelines to the dict
add_clips(clips_dict, timeline1.tracks[0])
add_clips(clips_dict, timeline2.tracks[0])

# Sort clips by their earliest timecode appearance across both timelines
clips_sorted = sorted(clips_dict.values(), key=lambda x: min(x.visible_range().start_time for x in [timeline1, timeline2] if x in clips_dict))

# Create a new timeline with a single track
new_timeline = otio.schema.Timeline(name="aggregated")
new_track = new_timeline.tracks.create(kind="Video")

# Add sorted clips to the new track
for clip in clips_sorted:
    new_track.append(clip)

# Write the result to a new .otio file
otio.adapters.write_to_file(new_timeline, "/path/to/aggregated.otio")
```

> Sorry, how would it do that? If a clip is present in multiple sequences on different tracks, at what track should it end up?

@BigRoy There are expected to be multiple variants of the same editorial project. For instance, Shot X, which requires multiple elements, will appear in multiple timelines with the same order and number of elements. However, there may be instances where additional clips with multiple elements are needed. In these cases, they will be added to new tracks while maintaining their vertical alignment with the main track.

**Comment by mkolar** [comment]
I agree that this should be revisited in a slightly different fashion. The value is clear, but the implementation should be handled probably outside of AYON.

I reckon let's separate out ayon-resolve from OP and revisit in there. @tweak-wtf I'm closing this as we need to clean the repo as much as possible of stale PRs, that are not targetin AYON, but please put a pin in it and let's continue the dicussion in it's own repo.
